client/clientservice/src/client_service.cpp

```diff
@@ -33,19 +33,24 @@ void ClientService::start(const std::string& addr, unsigned num_async_threads, u
     cqs_.emplace_back(builder.AddCompletionQueue());
   }
 
-  auto clientservice_server = std::unique_ptr<grpc::Server>(builder.BuildAndStart());
+  clientservice_server_ = std::unique_ptr<grpc::Server>(builder.BuildAndStart());
 
   // From the "C++ Performance Notes" in the gRPC documentation:
   // "Right now, the best performance trade-off is having numcpu's threads and one completion queue per thread."
   for (unsigned i = 0; i < num_async_threads; ++i) {
     server_threads_.emplace_back(std::thread([this, i] { this->handleRpcs(i); }));
   }
 
-  auto health = clientservice_server->GetHealthCheckService();
+  auto health = clientservice_server_->GetHealthCheckService();
   health->SetServingStatus(kRequestService, true);
   health->SetServingStatus(kEventService, true);
 
-  clientservice_server->Wait();
+  clientservice_server_->Wait();
+}
+
+void ClientService::shutdown() {
+  LOG_INFO(logger_, "Shutting down clientservice");
+  clientservice_server_->Shutdown();
```

Comment:
It looks like, in `client/clientservice/src/main.cpp`, we are starting the thread for signal handling, which can call `ClientService::shutdown()`, before we call `ClientService::start` (correct me if I'm misinterpreting here).

Do we not therefore need to handle the case where `ClientService::shutdown()` begins running before `ClientService::start` has made it to the line `clientservice_server_->Wait();`? (I do imagine this would likely have to involve someone interrupting the Client Service fairly soon after starting it, though it may also be technically possible if the scheduler arbitrarily delays the main thread somewhere between the point where the interrupt handler is set up and where ` clientservice_server_->Wait();` is reached). For starters, it looks like `clientservice_server_` could technically still be null here in that case. Furthermore, it looks like `ClientService::shutdown` uses `cqs_` and `server_threads_` but `ClientService::start` at some point before ` clientservice_server_->Wait();` modifies these.

Reply:
Thanks for the catch @upshaw-alex. I have updated the commit to fix the issue.
Note that since the call to `builder.BuildAndStart()` effectively starts the server, I decided to factor out the call to `wait` into another method called `wait`.

Reply:
Thank you for fixing this.